File: src/main/java/PdfGenerator.java
import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.font.PDType1Font;
import org.apache.pdfbox.pdmodel.font.Standard14Fonts;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.time.Duration;
import java.time.YearMonth;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;

class PdfGenerator {
    private static final float FONT_SIZE = 11f;
    private static final DateTimeFormatter MONTH_FORMATTER =
            DateTimeFormatter.ofPattern("MMMM yyyy", Locale.GERMAN);
    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm");

    static void writePdf(File output, YearMonth month, Employee employee, List<TimeEntry> days) throws IOException {
        byte[] template;
        try (InputStream is = PdfGenerator.class.getResourceAsStream("/stundenzettel-template.pdf")) {
            if (is == null) throw new IOException("PDF-Vorlage stundenzettel-template.pdf nicht im Klassenpfad gefunden");
            template = is.readAllBytes();
        }
        try (PDDocument doc = Loader.loadPDF(template)) {
            PDPage page = doc.getPage(0);
            PDType1Font font = new PDType1Font(Standard14Fonts.FontName.HELVETICA);
            try (PDPageContentStream cs = new PDPageContentStream(doc, page, PDPageContentStream.AppendMode.APPEND, true, true)) {
                cs.setFont(font, FONT_SIZE);
                drawText(cs, 285f, 715f, month.format(MONTH_FORMATTER));
                float employeeDetailsY = 686f;
                drawText(cs, 20f, employeeDetailsY, employee.lastName() + ", " + employee.firstName());
                drawText(cs, 162f, employeeDetailsY, employee.employeeId());
                drawText(cs, 310f, employeeDetailsY, employee.department());

                for (TimeEntry d : days) {
                    int day = d.date().getDayOfMonth();
                    float y = 584f - (day - 1) * 14.52f;
                    drawText(cs, 75f, y, d.start().format(TIME_FORMATTER));
                    drawText(cs, 176f, y, d.end().format(TIME_FORMATTER));
                    drawText(cs, 263f, y, formatDecimalHours(d.breakDuration()));
                    drawText(cs, 337f, y, formatDecimalHours(d.total()));
                    drawText(cs, 393f, y, d.remark());
                }

                Duration totalSum = days.stream().map(TimeEntry::total).reduce(Duration.ZERO, Duration::plus);
                drawText(cs, 337f, 134f, formatDecimalHours(totalSum));
            }
            Files.createDirectories(output.getAbsoluteFile().getParentFile().toPath());
            doc.save(output);
        }
    }

    static String formatDecimalHours(Duration d) {
        double hours = d.toSeconds() / 3600.0;
        return String.format(Locale.GERMAN, "%.2f", hours);
    }

    private static void drawText(PDPageContentStream cs, float x, float y, String text) throws IOException {
        cs.beginText();
        cs.newLineAtOffset(x, y);
        cs.showText(text);
        cs.endText();
    }
}

File: src/lib/pdfGenerator.ts
import { PDFDocument, type PDFPage, type PDFFont } from "pdf-lib";
import fontkit from "@pdf-lib/fontkit";
import { Duration, DateTimeFormatter } from "@js-joda/core";
import { Effect } from "effect";
import type { Timesheet } from "./xslxParser";
import { GERMAN_MONTHS } from "./months";
const TEMPLATE_URL = "/stundenzettel-vorlage.pdf";
const FONT_URL = "/fonts/JetBrainsMono-Regular.ttf";

const LAYOUT = {
    fontSize: 10,
    yearMonth: { x: 285, y: 715 },
    employeeDetails: {
        fullName: { x: 20, y: 686 },
        employeeId: { x: 162, y: 686 },
        department: { x: 310, y: 686 },
    },
    table: {
        firstRowY: 584.25,
        rowHeight: 14.52,
        columns: {
            start: 76,
            finish: 177,
            break: 263,
            workedHours: 337,
            remark: 393,
        },
    },
    totalWorkedHours: { x: 337, y: 134 },
} as const;

const TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm");


export function generatePdf(timesheet: Timesheet, year: number, month: number): Effect.Effect<Uint8Array, Error> {
    return Effect.gen(function* () {
        const [templateBytes, fontBytes] = yield* Effect.all([
            fetchArrayBuffer(TEMPLATE_URL, `PDF-Vorlage "${TEMPLATE_URL}" konnte nicht geladen werden`),
            fetchArrayBuffer(FONT_URL, `Schriftart "JetBrains Mono" konnte nicht geladen werden`),
        ], { concurrency: "unbounded" });

        return yield* Effect.tryPromise({
            try: () => stampTimesheet(templateBytes, fontBytes, timesheet, year, month),
            catch: (e) => new Error(`PDF konnte nicht erzeugt werden: ${e}`),
        });
    });
}

function fetchArrayBuffer(url: string, errorMessage: string): Effect.Effect<ArrayBuffer, Error> {
    return Effect.tryPromise({
        try: () => fetch(url).then((response) => {
            if (!response.ok) throw new Error(`Status ${response.status}`);
            return response.arrayBuffer();
        }),
        catch: (e) => new Error(`${errorMessage}: ${e}`),
    });
}

async function stampTimesheet(templateBytes: ArrayBuffer, fontBytes: ArrayBuffer, timesheet: Timesheet, year: number, month: number): Promise<Uint8Array> {
    const pdf = await PDFDocument.load(templateBytes);
    pdf.registerFontkit(fontkit);
    const font = await pdf.embedFont(fontBytes, { subset: true });
    const page = pdf.getPages()[0];

    const draw = drawTextAt(page, font);

    draw(`${GERMAN_MONTHS[month - 1]} ${year}`, LAYOUT.yearMonth.x, LAYOUT.yearMonth.y);

    const { firstName, lastName, employeeId, department } = timesheet.employeeDetails;
    draw(`${lastName}, ${firstName}`, LAYOUT.employeeDetails.fullName.x, LAYOUT.employeeDetails.fullName.y);
    draw(employeeId, LAYOUT.employeeDetails.employeeId.x, LAYOUT.employeeDetails.employeeId.y);
    draw(department, LAYOUT.employeeDetails.department.x, LAYOUT.employeeDetails.department.y);

    const { firstRowY, rowHeight, columns } = LAYOUT.table;
    let totalWorked = Duration.ZERO;
    for (const entry of timesheet.entries) {
        const y = firstRowY - (entry.date.dayOfMonth() - 1) * rowHeight;
        draw(entry.start.format(TIME_FORMAT), columns.start, y);
        draw(entry.finish.format(TIME_FORMAT), columns.finish, y);
        draw(formatDecimalHours(entry.break), columns.break, y);
        draw(formatDecimalHours(entry.workedHours), columns.workedHours, y);
        const remark = entry.remark.length > 18 ? entry.remark.slice(0, 18) + "..." : entry.remark;
        draw(remark, columns.remark, y);
        totalWorked = totalWorked.plus(entry.workedHours);
    }

    draw(formatDecimalHours(totalWorked), LAYOUT.totalWorkedHours.x, LAYOUT.totalWorkedHours.y);

    return pdf.save();
}

function drawTextAt(page: PDFPage, font: PDFFont) {
    return (text: string, x: number, y: number) =>
        page.drawText(text, { x, y, size: LAYOUT.fontSize, font });
}

function formatDecimalHours(duration: Duration): string {
    const hours = duration.toMillis() / 3_600_000;
    return hours.toFixed(2).replace(".", ",");
}
